token/program-2022/src/extension/mod.rs

```diff
@@ -1124,12 +1125,51 @@ impl Extension for AccountPaddingTest {
     const TYPE: ExtensionType = ExtensionType::AccountPaddingTest;
 }
 
+/// Packs an unsized extension into a new TLV space
+///
+/// This function reallocates the account as needed to accommodate for the
+/// change in space, then allocates in the TLV buffer, and finally writes the
+/// bytes.
+pub fn alloc_and_serialize<S: BaseState, V: UnsizedExtension>(
+    account_info: &AccountInfo,
+    value_bytes: &[u8],
```

Comment:
I wonder if we could lock this down a little, rather than accepting any arbitrary bytes. But like you alluded to [here](https://github.com/solana-labs/solana-program-library/pull/4704), I too don't love the idea of being tied to borsh for serialization of UnsizedExtensions.
So I don't know if it's worth the trouble, but I'm imagining something like:
```
// maybe this could just be in an SPL library instead of waiting on solana-program
trait UnsizedPack {
    fn pack_into_slice(&self, dst: &mut [u8]) -> Result<(), ProgramError>;
    fn unpack_from_slice(src: &[u8]) -> Result<Self, ProgramError>;
    fn get_packed_len() -> usize;

    fn pack(&self, dst: &mut [u8]) -> Result<(), ProgramError> {
        if dst.len() != self.get_packed_len() {
            return Err(ProgramError::InvalidAccountData);
        }
        self.pack_into_slice(dst);
        Ok(())
    }
}
```

```
trait UnsizedExtension: Extension + UnsizedPack {}

pub fn alloc_and_serialize<S: BaseState, V: UnsizedExtension(
    account_info: &AccountInfo,
    extension: V,
) -> Result<(), ProgramError> {
    ...
    let mut state = StateWithExtensionsMut::<S>::unpack(&mut buffer)?;
    let data = state.alloc::<V>(extension.get_packed_len(), false)?;
    extension.pack(data);
}
```

Then the TokenMetadata extension/interface could implement something like:
```
fn pack_into_slice(&self, dst: &mut [u8]) -> Result<(), ProgramError> {
    let bytes = self.try_to_vec()?;
    dst.copy_from_slice(value_bytes);
}
```

Reply:
I like this approach! I put in #4719 to add that trait to the TLV library so that it can be used here

Reply:
And I put in #4720, thanks so much for the suggestion! Once that lands, I can rebase this PR and just use `UnsizedPack` directly. I'll aim to do it in new commits to make this a bit easier for re-review